drivers/vmwarevsphere/utils.go

```diff
@@ -35,6 +38,122 @@ func (d *Driver) remoteExec(procman *guest.ProcessManager, arg string) (int64, e
 	return code, nil
 }
 
+func (d *Driver) getDatastore(spec *types.VirtualMachineConfigSpec) (*object.Datastore, error) {
+	var sp *object.StoragePod
+	if d.DatastoreCluster != "" {
+		log.Infof("Finding datastore cluster %s", d.DatastoreCluster)
+		sp, err := d.finder.DatastoreCluster(d.getCtx(), d.DatastoreCluster)
+		if err != nil {
+			return nil, err
+		}
+
+		return d.recommendDatastore(sp, spec)
+	}
+
+	if d.Datastore != "" {
+		log.Infof("Finding datastore %s", d.Datastore)
+		return d.finder.Datastore(d.getCtx(), d.Datastore)
+	}
+
+	//nothing set, try default ds cluster then default ds
+	log.Infof("Finding default datastore cluster")
+	sp, err := d.finder.DefaultDatastoreCluster(d.getCtx())
+	if err != nil {
+		return nil, err
+	}
+
+	if sp != nil {
+		return d.recommendDatastore(sp, spec)
+	}
+
+	log.Infof("No default datastore cluster, finding default datastore")
+	return d.finder.DefaultDatastore(d.getCtx())
+}
+
+func (d *Driver) recommendDatastore(sp *object.StoragePod, spec *types.VirtualMachineConfigSpec) (*object.Datastore, error) {
+	// Build pod selection spec from config spec
+	spref := sp.Folder.Reference()
+	podSelectionSpec := types.StorageDrsPodSelectionSpec{
+		StoragePod: &spref,
+	}
+
+	// Keep list of disks that need to be placed
+	var disks []*types.VirtualDisk
+	// Collect disks eligible for placement
+	for _, deviceConfigSpec := range spec.DeviceChange {
+		s := deviceConfigSpec.GetVirtualDeviceConfigSpec()
+		if s.Operation != types.VirtualDeviceConfigSpecOperationAdd {
+			continue
+		}
+
+		if s.FileOperation != types.VirtualDeviceConfigSpecFileOperationCreate {
+			continue
+		}
+
+		d, ok := s.Device.(*types.VirtualDisk)
+		if !ok {
+			continue
+		}
+
+		podConfigForPlacement := types.VmPodConfigForPlacement{
+			StoragePod: spref,
+			Disk: []types.PodDiskLocator{
+				{
+					DiskId:          d.Key,
+					DiskBackingInfo: d.Backing,
+				},
+			},
+		}
+
+		podSelectionSpec.InitialVmConfig = append(podSelectionSpec.InitialVmConfig, podConfigForPlacement)
+		disks = append(disks, d)
+	}
+
+	rp := d.resourcepool.Reference()
+	sps := types.StoragePlacementSpec{
+		Type:             string(types.StoragePlacementSpecPlacementTypeCreate),
+		ResourcePool:     &rp,
+		PodSelectionSpec: podSelectionSpec,
+		ConfigSpec:       spec,
+	}
+
+	c, err := d.getSoapClient()
```

Comment:
nit: do you think we can move lines 120-122 to the beginning of the method? So in case error happens, we can save on `disks` population logic

Reply:
the client is used all over and i'd rather refactor to having it on the driver struct and just doing d.client everywhere. all the stuff is spec creation and the loop is usually done once, wont save any one much and if i move the client to the struct this is moot